openhcl/fixed_pool_alloc/src/lib.rs

```diff
@@ -0,0 +1,530 @@
+            let linear = MemoryRange::from_4k_gpn_range(std::ops::Range {
+                start: chunk.base_pfn,
+                end: chunk.base_pfn + chunk.size_pages,
+            });
+            for range in fixed_pool {
+                if range.contains(&linear) {
+                    if chunk.allocated {
+                        pages.push(State::Restored {
+                            base_pfn: chunk.base_pfn,
+                            size_pages: chunk.size_pages,
+                            tag: chunk.tag.clone(),
+                        });
+                    } else {
+                        pages.push(State::Free {
+                            base_pfn: chunk.base_pfn,
+                            size_pages: chunk.size_pages,
+                        });
+                    }
+                }
+            }
+        });
+
+        Ok(Self {
+            inner: Arc::new(Mutex::new(FixedPoolInner { state: pages })),
+        })
+    }
+
+    /// Validate memory pool after restore finishes.
+    pub fn validate(&self) -> anyhow::Result<(), FixedPoolIntegrity> {
+        let inner = self.inner.lock();
+        let leaked_blocks = inner
+            .state
+            .iter()
+            .filter(|chunk| matches!(chunk, State::Restored { .. }))
+            .count();
+
+        if leaked_blocks > 0 {
+            return Err(FixedPoolIntegrity { leaked_blocks });
+        }
+
+        Ok(())
+    }
+
+    /// Return an allocator instance that can be used to allocate pages.
+    pub fn allocator(&self) -> FixedPoolAllocator {
+        FixedPoolAllocator {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+/// A page allocator for fixed memory.
+///
+/// Pages are allocated via the [`Self::alloc`] method and freed by dropping the
+/// associated handle returned.
+#[derive(Clone, Debug)]
+pub struct FixedPoolAllocator {
+    inner: Arc<Mutex<FixedPoolInner>>,
+}
+
+impl FixedPoolAllocator {
+    const VFIO_MSHV_TAG: &str = "mshv_dma";
+
+    /// Allocate contiguous pages from the fixed pool with the given
+    /// tag. If a contiguous region of free pages is not available, then an
+    /// error is returned.
+    pub fn alloc(
+        &self,
+        size_pages: NonZeroU64,
+        tag: String,
+    ) -> Result<FixedPoolHandle, FixedPoolOutOfMemory> {
+        let mut inner = self.inner.lock();
+        let size_pages = size_pages.get();
+
+        let index = inner
+            .state
+            .iter()
+            .position(|state| match state {
+                State::Free {
+                    base_pfn: _,
+                    size_pages: len,
+                } => *len >= size_pages,
+                State::Allocated { .. } | State::Restored { .. } | State::Confirmed { .. } => false,
+            })
+            .ok_or(FixedPoolOutOfMemory {
+                size: size_pages,
+                tag: tag.clone(),
+            })?;
+
+        let base_pfn = match inner.state.swap_remove(index) {
```

Comment:
It is unreachable because the check is already happened at the line 306 - e.g. position() call already returned the index of Free block and the index must be valid, any other block states must panic.

Strictly speaking, this is double check which may not be necessary, but existing shared_pool_allocator does same, so I decided to keep it.